### ISSUE TITLE
[C#] Implement new Concept Exercise: structs

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1017]
This issue describes how to implement the `structs` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the Concept of Structs in C#.

## Learning objectives

- Know what structs are.
- Know how to define a `struct`.
- Know how to add members to structs.
- Know the differences between structs and classes.

## Out of scope

- ref structs.
- ref returns
- ref locals
- Struct finalizers.
- `Span<T>`/`Memory<T>`
- `stackalloc`

## Concepts

This Concepts Exercise's Concepts are:

- `structs`: know what structs are; know how to define a `struct`; know how to add members to structs; know the differences between structs and classes.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `inheritance`: know that all types are derived from `object`.
- `classes`: know how to define and work with classes.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

### Hints

- [structs][structs]: introduction to structs.

### After

- [structs][structs]: introduction to structs.
- [class-or-struct][class-or-struct]: lists the guidelines for choosing between a struct and class.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/types/struct.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[structs]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct
[class-or-struct]: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

**Comment by mikedamay** [comment]
We should mention `struct`s' support for interfaces either in _after.md_ here or somewhere else in the concept exercises

**Comment by mikedamay** [comment]
I think this exercise should cover equality as it relates to `struct`s.  See https://github.com/exercism/v3/issues/1016#issuecomment-640538913

**Comment by mikedamay** [comment]
> Know how to add members to structs.

I'm not sure what that means.

**Comment by mikedamay** [comment]
> Know the differences between structs and classes.

This breaks down into:
1. Demonstrate passing by value
2. Demonstrate assignment of one struct to another
3. Demonstrate that it is not possible for a struct to be a field of another type and to have guaranteed valid arbitrary values
4. Equality tests

Equality (4) and Assignment (2) are straightforward.  

Passing by value (1) is a hard thing to show given our usual template format/puzzle type - i.e. either providing a method signature or inviting the student to completely implement a method.  We could do this if we allowed the approach of providing a broken implementation and asking the student to fix it.  On the other hand maybe just having a method with a ref type would be sufficient.  Meh, maybe not.

The absence of a default constructor (3) and its consequences are also tough to get across although I suppose there could be a method to create a valid object (containing a struct) but I think much of the point would be lost.  Again, in the absence of a wider range of puzzle types we will have to be very imaginative (i.e. more imaginative than me) to make our point with the student.

**Comment by ErikSchierboom** [comment]
>> Know how to add members to structs.
> I'm not sure what that means. Do you mean change the definition?

Oh, I just mean that we should explain that structs can also have methods etc., just like classes.

**Comment by ErikSchierboom** [comment]
> We should mention structs' support for interfaces either in after.md here or somewhere else in the concept exercises

We have an interfaces exercise already, we could either add it there or here. Do we want this to depend on `interfaces`, or do we want `interfaces` to depend on `structs`? My preference would be with the former, but I can be convinced either way.

**Comment by mikedamay** [comment]
Agreed structs should have interfaces as a prerequisite.  Do we want to have an interface puzzzle as part of this exercise or just mention it in `after.md`?

**Comment by ErikSchierboom** [comment]
> Passing by value (1) is a hard thing to show given our usual template format/puzzle type - i.e. either providing a method signature or inviting the student to completely implement a method. We could do this if we allowed the approach of providing a broken implementation and asking the student to fix it. On the other hand maybe just having a method with a ref type would be sufficient. Meh, maybe not.

I think that this is actually something that I decided against doing later. It might make far more sense to explain this when explaining `ref` parameters. 

> The absence of a default constructor (3) and its consequences are also tough to get across although I suppose there could be a method to create a valid object (containing a struct) but I think much of the point would be lost. Again, in the absence of a wider range of puzzle types we will have to be very imaginative (i.e. more imaginative than me) to make our point with the student.

The absence of a default constructor? But structs always have a default constructor? This might be something that we just mention and not make part of the exercise.

**Comment by ErikSchierboom** [comment]
> Agreed structs should have interfaces as a prerequisite. Do we want to have an interface puzzzle as part of this exercise or just mention it in after.md?

I'd say: just mention it in the `after.md` file.

**Comment by ErikSchierboom** [comment]
For some inspiration, maybe also take a look at [F#'s records exercise](https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/records).

**Comment by mikedamay** [comment]
Please consider the [instructions](https://github.com/mikedamay/v3/blob/csharp/structs/languages/csharp/exercises/concept/structs/.docs/instructions.md) and [code](https://github.com/mikedamay/v3/blob/csharp/structs/languages/csharp/exercises/concept/structs/.meta/Example.cs) for this exercise.

The story depends on speculators claiming plots of land based on their position, shape and dimensions.  It's not really  clear why these plots of land haven't got an identifier.  I will check with Elixir to see if they came up with an idea that fits and isn't flawed but assuming not, is this OK?

**Comment by ErikSchierboom** [comment]
Excellent work! Really nice exercise.

**Comment by mikedamay** [comment]
Closed by PR #1857
